feat: add `aria-required` attributes to WC form fields

frosso opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adding the `aria-required="true"` attribute to required form inputs.
I chose to add the `aria-required` attribute rather than the `required` attribute to avoid backwards-compatibility issues.
When the `required` attribute is added, the form submission is intercepted and prevented by the browser.
By adding `aria-required`, instead`, we add information to Screen Reader users without affecting the form's behavior.

I ensured that the `aria-required` attribute was added to all input types, excluding `hidden`.
The `aria-required` attribute is added to all inputs of type `radio`, [per HTML specs](https://html.spec.whatwg.org/multipage/input.html#the-required-attribute):
> To avoid confusion as to whether a [radio button group](https://html.spec.whatwg.org/multipage/input.html#radio-button-group) is required or not, authors are encouraged to specify the attribute on all the radio buttons in a group. 

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes https://github.com/woocommerce/woocommerce/issues/44663

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

- (optional) Add the following snippet to your installation to test the radio inputs(you can use [the Code Snippets plugin](https://wordpress.org/plugins/code-snippets/), for example)
```php
add_action( 'woocommerce_after_checkout_billing_form', 'checkout_test_radio_options' );
function checkout_test_radio_options() {
	echo "<div>";
	echo sprintf( '<p>%s</p>', __( "Test radio input" ) );
    woocommerce_form_field( 'test_radio',  array(
		'required'  => true,
        'type'      => 'radio',
        'class'     => array( 'form-row-wide' ),
        'options'   => array(
            '1' => 'Option 1',
            '2' => 'Option 2',
        ),
    ), '1' );
	echo "</div>";
}
```
- Add a product to the cart
- Go to the shortcode checkout page
- The checkout page should display a wide range of inputs. All required inputs (excluding the "hidden" ones) should have an `aria-required="true"` attribute

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [x] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
feat: add `aria-required` attributes to WC form fields
</details>


<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

frosso committed:
feat: add `aria-required` attributes to WC form fields

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

frosso committed:
added tests for radio fields

frosso committed:
WIP

frosso committed:
WIP

frosso committed:
yoda, space

frosso committed:
Update functions.php

frosso committed:
WIP

frosso committed:
WIP

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

frosso commented:
Assigned to Rubik because the change affects checkout form fields on classic checkout. Feel free to reassign if incorrect!

alexflorisca commented on plugins/woocommerce/tests/legacy/unit-tests/templates/functions.php:
```diff
@@ -193,4 +193,49 @@ public function test_hidden_field() {
+			'test',
+			array(
+				'type'     => 'radio',
+				'id'       => 'test',
```
This should make the test pass
```suggestion
				'id'       => 'test_radio',
```

frosso replied:
Ope, thank you!

alexflorisca reviewed (commented):
The code looks good and the `aria-required` attribute is added for all inputs except hidden ones. However for the `checkbox` input type, the `aria-label` is added onto the `label` rather than the `input`. 

The test you added is also failing, I added a suggestion below that should fix it. 

Nice work, ping me again when you see this and push changes and I'll have another look 👍

frosso committed:
Update plugins/woocommerce/tests/legacy/unit-tests/templates/functions.php

Co-authored-by: Alex Florisca <alex.florisca@automattic.com>

frosso committed:
fix checkbox required attribute, add tests

frosso committed:
lint

frosso commented:
Thank you @alexflorisca ! I fixed the linting issues & added a new test for checkboxes (ensuring the `aria-attribute` is added to the input).
Thanks again for your time!

opr reviewed (changes requested):
This is a really nice clean solution, and I like that you opted for `aria-required`.

The changes you made here won't apply to country or state select fields as these are converted into select2 instances. For these you would need to update `selectWoo.full.js`.

https://github.com/woocommerce/woocommerce/commit/bbabe8a8efab6a74633062b6955f634c74c559f1 shows how this would look, though you need to change the `country-select.js` code to check for `aria-required` instead of `required`

Also in 718bd3c9800efa9a63e680862ec5f421c559caab I made it so the asterisk is not read by screen readers. What do you think about applying this change too?

frosso commented:
Thank you for the feedback, @opr !

> The changes you made here won't apply to country or state select fields as these are converted into select2 instances. For these you would need to update `selectWoo.full.js`.

> https://github.com/woocommerce/woocommerce/commit/bbabe8a8efab6a74633062b6955f634c74c559f1 shows how this would look, though you need to change the `country-select.js` code to check for `aria-required` instead of `required`

With this approach, were you able to make the Screen Reader (SR) software announce the field as "required"?
With the changes implemented (as a note - slightly differently because the `$this.attr('required')` doesn't return a boolean but a string & applying the `required` attribute on the `role="combobox"` element), I am still not able to have any software announce the field as required.
I tried with VoiceOver on iOS/Mac OS, Talkback on Android Chrome & Brave, [Screen Reader in Chrome](https://chromewebstore.google.com/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn?pli=1).

> Also in 718bd3c I made it so the asterisk is not read by screen readers. What do you think about applying this change too?

I'd like to hear your thoughts on the issue I mentioned above above, first.
Since some browsers & SR software combinations don't announce required fields (e.g.: [Talkback on Android w/ Google Chrome](https://www.tpgi.com/required-attribute-requirements/)), I think the "required" `abbr` (albeit redundant in some other cases) provides a useful hint to users that otherwise wouldn't get the hint.

If we removed the `required` abbr, do you think we should also remove the `(optional)` hint on optional fields at checkout?

FWIW, I also noticed that to be really consistent, we'd need to add a `aria-hidden` attribute [on `address-i18n.js` as well](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce/client/legacy/js/frontend/address-i18n.js#L17-L21).

Since both issues go hand-in-hand, I was instead wondering if we should keep the "required" `abbr` tag to ensure that required `select2` fields can be announced by SR software.

opr commented:
@frosso 

> With this approach, were you able to make the Screen Reader (SR) software announce the field as "required"?
With the changes implemented (as a note - slightly differently because the $this.attr('required') doesn't return a boolean but a string & applying the required attribute on the role="combobox" element), I am still not able to have any software announce the field as required.
I tried with VoiceOver on iOS/Mac OS, Talkback on Android Chrome & Brave, [Screen Reader in Chrome](https://chromewebstore.google.com/detail/screen-reader/kgejglhpjiefppelpmljglcjbhoiplfn?pli=1).

Yes, I was, I just implemented my changes on your branch. I tried with VoiceOver and Screen reader in Chrome and both worked for me! Can you check out the [`origin/try/country-screen-reader` branch](https://github.com/woocommerce/woocommerce/tree/try/country-screen-reader) and see if it works there?

> Since some browsers & SR software combinations don't announce required fields (e.g.: [Talkback on Android w/ Google Chrome](https://www.tpgi.com/required-attribute-requirements/)), I think the "required" abbr (albeit redundant in some other cases) provides a useful hint to users that otherwise wouldn't get the hint.

I didn't realise that some wouldn't announce the field as required! I agree the hints are probably useful then, let's leave them!

frosso commented:
> Yes, I was, I just implemented my changes on your branch. I tried with VoiceOver and Screen reader in Chrome and both worked for me! Can you check out the [origin/try/country-screen-reader branch](https://github.com/woocommerce/woocommerce/tree/try/country-screen-reader?rgh-link-date=2024-06-17T13%3A39%3A37Z) and see if it works there?

I applied those changes on my local environment.
I also added a required `select` on checkout with label "Test required input".

I couldn't record Talkback on Android, due to some permissions issue. However, this is the transcribed result on Brave browser:

<details>
<summary>Brave Browser transcription</summary>
"Test required input" label:
> Test required input

"Test required input" input:
> Collapsed, required, test required input, menu popup button, double tap to activate

"First name" label:
> First name

"First name" input:
> Required, edit box, first name, double tap to edit text

"Country / region" label
> Country / region

"Country / region" selectWoo input
> Collapsed, United States US, dropdown list, country / region, double tap to change
</details>

On Android with Talkback & Google Chrome, all inputs (and labels) are not announced as required.

Mac OS w/ VoiceOver on Safari - VoiceOver doesn't seem to announce the "selectWoo" inputs as required. It's skipping the "required" `abbr` tag and announcing other native fields as `required`.

https://github.com/woocommerce/woocommerce/assets/273592/00292fa4-dd3b-48a2-931c-3070e9fa5f40

So with that branch's changes the "Country / Region" was never announced as required during my testing.

opr commented:
@frosso very strange! It looks like that select isn't styled correctly , I'm wondering if it really is being picked up by select2 on your site?

Does it have `aria-required="true"` on it?

When I inspect it I see this, what do you see?

<img width="387" alt="image" src="https://github.com/woocommerce/woocommerce/assets/5656702/cf35d5a9-c61f-4aa2-854b-6683e36cb9ef">

frosso commented:
> Does it have aria-required="true" on it?
> When I inspect it I see this, what do you see?

I have a similar output
<img width="1861" alt="Screenshot 2024-06-17 at 5 18 51 PM" src="https://github.com/woocommerce/woocommerce/assets/273592/99a622bd-6f2d-41ab-84fb-2841eb4064f2">

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

frosso committed:
importing changes from Thomas' branch, aria-hidden on abbr

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

frosso committed:
added hidden also on JS changes

frosso commented:
@opr thank you for the time you dedicated today. As we discussed, I made the following changes:
- Added the `aria-required` to the combobox/selectWoo
- Added `aria-hidden` on the `required` abbr

frosso committed:
template number change

opr reviewed (changes requested):
Hi - it looks like the changes to selectWoo.full.js aren't copied over so the country doesn't have the `aria-required` attribute.

frosso committed:
forgot to commit these changes

frosso commented:
@opr 🤦 I had it modified locally and didn't push it

Do you happen to know if I also need to modify `plugins/woocommerce/client/legacy/js/selectWoo/selectWoo.js` somehow? Or if I need to implement the changes in https://github.com/woocommerce/selectWoo?

opr commented:
> @opr 🤦 I had it modified locally and didn't push it
> 
> Do you happen to know if I also need to modify `plugins/woocommerce/client/legacy/js/selectWoo/selectWoo.js` somehow? Or if I need to implement the changes in [woocommerce/selectWoo](https://github.com/woocommerce/selectWoo)?

I couldn't see any code path that loaded `plugins/woocommerce/client/legacy/js/selectWoo/selectWoo.js`  but maybe you want to do it just in case, with that said I didn't know about that selectWoo repository, makes sense that we should edit there as well. @barryhughes could you shed some light on the repository @frosso linked, is this the actual source for selectWoo and should we be editing the files there instead?

barryhughes commented:
> @barryhughes could you shed some light on the repository @frosso linked, is this the actual source for selectWoo and should we be editing the files there instead?

Yes, in theory that is the actual source of SelectWoo. Unfortunately, I believe there may have been some divergence over time, with some edits being made directly to the copy *here,* but not always being applied over there (and possibly vice versa).

If we can also raise a PR making the same change over there, though, I think that would be beneficial. cc @naman03malhotra for any other thoughts, as you've recently been looking at Select2/Woo.

naman03malhotra commented:
>, I believe there may have been some divergence over time, with some edits being made directly to the copy here, but not always being applied over there (and possibly vice versa).

If we were to make changes in the WooSelect repo and override the build file in core, we would lose the changes we make directly to the copy in core. So, we would need to do some further deep diving to make sure we have a common set of changes in the final result.

naman03malhotra reviewed (commented):
>, I believe there may have been some divergence over time, with some edits being made directly to the copy here, but not always being applied over there (and possibly vice versa).

If we were to make changes in the WooSelect repo and override the build file in core, we would lose the changes we made directly to the copy in core. So, we would need to do some further deep diving to make sure we have a common set of changes in the final result and that maybe a tricky thing to do.

frosso commented:
@barryhughes @naman03malhotra I just took a quick look - after all the moving around, [these are the latest changes](https://github.com/woocommerce/woocommerce/commits/37205ff4e0f489f28345895b12a0d5d602b0af4d/assets/js/selectWoo) to selectWoo in this repository.
Notice [the last commit](https://github.com/woocommerce/woocommerce/commit/41f63bd4877c5fd4f808b77eb7693ef08bc8c2d1) matches [the last commit in the selectWoo repository](https://github.com/woocommerce/selectWoo/commit/69d51b0624b2925a3c121d78b8e5d15d45bdb955) 3 years ago.

I created this PR on the selectWoo repository to basically mirror these changes: https://github.com/woocommerce/selectWoo/pull/53

frosso committed:
porting of changes from selectWoo repository

frosso committed:
undo unwanted changes

opr committed:
Update checkout e2e to stop looking for * in field labels

barryhughes commented:
Thanks, @frosso: @naman03malhotra I assigned to you (the usual round-robin doesn't seem to work in that repo, and I have low availability next week so couldn't grab it myself).

frosso committed:
removed aria-hidden to abbr, to avoid regressions across multiple repos

frosso committed:
lint

frosso commented:
The addition of `aria-hidden` on the `abbr` would cause regressions across multiple repositories using e2e tests that run on multiple WC core versions.
We decided to defer this as a future enhancement: https://github.com/woocommerce/woocommerce/issues/48948

barryhughes commented:
@frosso should https://github.com/woocommerce/selectWoo/pull/53 also be put on hold?

frosso commented:
> @frosso should https://github.com/woocommerce/selectWoo/pull/53 also be put on hold?

@barryhughes sorry I missed your comment - I don't think there's anything that should hold it

naman03malhotra reviewed (commented):
>I don't think there's anything that should hold it

@frosso I am still able to see the changed made to `select2` files, in this PR https://github.com/woocommerce/selectWoo/pull/53

Since we have now [removed](https://github.com/woocommerce/woocommerce/pull/48731) Select2 altother from core, probably we can also remove it from the SelectWoo repo to avoid confusion.

naman03malhotra reviewed (approved):
Thank you for working on it, works well with the given testing instructions for checkout shortcode page.

<img width="874" alt="image" src="https://github.com/user-attachments/assets/a80f1830-29c8-44fd-8f9f-fd7e77255623">

naman03malhotra reviewed (commented):
>Do you think this PR is currently on hold because select2 is not currently removed from the selectWoo repository?

Sorry for the confusion, I though [that PR](https://github.com/woocommerce/selectWoo/pull/53) and this are meant to go together, but I see they have separate concerns.

>as it stands, the selectWoo repository still contains Select2.
How would you prefer to proceed? I have a few options:
First remove Select2 from [the SelectWoo repository](https://github.com/woocommerce/selectWoo), then update https://github.com/woocommerce/selectWoo/pull/53 accordingly
First merge https://github.com/woocommerce/selectWoo/pull/53 , then update the SelectWoo repository to remove select2
I made the changes to the selectWoo repository just as a mirror of the changes in the main WC Core repository.

Either approach looks good to me, I'll let you decide how to proceed. https://github.com/woocommerce/selectWoo/pull/53 Looks good aparart from minor changes I've suggested.

opr reviewed (approved):
Re-tested, and it's working well for me, thanks @frosso I think rebasing should fix the issues with failing CI.

frosso committed:
Merge branch 'trunk' into frosso-patch-1
